Add documentation about the Zuul jobs (#567)
infrawatch/service-telemetry-operator: ci/README.md

# service-telemetry-operator CI playbooks

The playbooks in this directory are used by zuul jobs, which are defined in ../.zuul.yaml.

## Job descriptions

### PR jobs

There are 6 jobs run on every PR that is targeting master.
These are reported under the `rdoproject.org/github-check` check.

Two scenarios run:
- `local_build`, which replaces build the STF images and deploys by creating a STF object.
- `local_build-index_deploy`, which builds the images and does an index-based deployment

Each of these scenarios run across the following OCP versions:
- 4.12
- 4.13
- 4.14

> ### Periodic jobs

Reviewer: It will be good to mention that periodic jobs is being run daily.

Author: This is mentioned as "jobs are run nightly"

Below are examples of how to add a job. Take note of how the `scenario` var and the `nodeset` is passed.

    - job:
        name: stf-crc-nightly_bundles
        parent: stf-base
        abstract: true
        description: |
          Example of a job that extends the stf-base-job, and passes the "nightly_bundles" scenario var. This job does NOT have a nodeset defined so it must be abstract.
        vars:
          scenario: "nightly_bundles"

OR

    - job:
        name: stf-crc-ocp_414-nightly_bundles
        parent: stf-crc-nightly_bundles
        description: |
          Example of a job defining a nodeset to be used.
          Since this job derives from a job with a scenario, it can be run directly.
        nodeset: stf-crc_extracted-ocp414

All non-abstract jobs inheriting from `stf-base` must pass a `scenario` var to work correctly. There is no default value for the `scenario`.
All non-abstract jobs defined in this repo must have a `nodeset` to run correctly. Specifically, the nodeset must include nodes called `controller` and `crc`. This requirements comes from the `stf-base-2node` job.

> 

Reviewer: We need to mention that after define the job it should be added to the project template or under the project itself to be fired.

## Troubleshooting

## FAQ

> 

Reviewer: we should mention the github configurations RDO for the infrawatch projects. and Mention how this repo is being connected to this configuration so we are able to use the pipelines and also the node-set which already configured there. https://review.rdoproject.org/cgit/config/tree/zuul/rdo.yaml?id=77800d05c96e6378819d074ea47bb3965c0b8905#n1255

## How to use zuul in an new repo ?

Author: Excellent, I was hoping for some feedback on this section.

The OCP version selection is done by specifying the `nodeset` to be used by the job.
The `nodesets` are defined in `.zuul.yaml`. Each nodeset corresponds to a different version of OCP.
Each nodeset contains two hosts: `crc` and `controller`.
All ansible playbooks are run against `controller`.

The rest of this section provides further detail on the OCP version selection, and how it relates to CRC and the deployment topology.

The nodesets select the hosts based on labels in zuul.
The labels available in zuul are shown on the [RDO Zuul labels tab](https://review.rdoproject.org/zuul/labels).

The labels used for the nodesets are `coreos-crc-extracted-<crc_version>-<size>`.
The “extracted” CRC describes the way that the job deploys and interacts with CRC.

Usually, CRC is run using the `crc start` command, which created a VM on your host which runs the OCP cloud.
In Zuul, the provisioned hosts are also virtual machines, so running `crc start` would result in a VM in a VM. This nested virtualisation causes some performance issues.

The `extracted` deployment try to address the performance issues associated with nested virtualisation. The infrastructure is more complicated than nested.

> The `coreos-crc-extracted-...` labels provide a VM with an extracted CRC VM image, so that the CRC VM can be booted directly by the cloud provider. The `crc` VM is not accessed directly, but via a second `controller` VM, on which tests are run. The `stf-base-2node` job includes a network configuration to make sure the controller can communicate with the OCP deployment in CRC. All ansible playbooks should be run against `controller`. The `stf-base-2node` job is a common parent to all the jobs in the repo, and should rarely need an update.

Reviewer 2: NIT "All ansible playbooks should be run against `controller`. The `stf-base-2node` job is a common parent to all the jobs in the repo, and should rarely need an update." is already mentioned in L37 and L54

> The `coreos-crc-extracted-...` labels provide a VM with an extracted CRC VM image, so that the CRC VM can be booted directly by the cloud provider. The `crc` VM is not accessed directly, but via a second `controller` VM, on which tests are run. The `stf-base-2node` job includes a network configuration to make sure the controller can communicate with the OCP deployment in CRC. All ansible playbooks should be run against `controller`. The `stf-base-2node` job is a common parent to all the jobs in the repo, and should rarely need an update.

Reviewer 3: ```suggestion
The `coreos-crc-extracted-...` labels provide a VM with an extracted CRC VM image, so that the CRC VM can be booted directly by the cloud provider. The `crc` VM is not accessed directly, but via a second `controller` VM, on which tests are run. The `stf-base-2node` job includes a network configuration to make sure the controller can communicate with the OCP deployment in CRC.
```

Those last two sentences seem a bit off topic in this paragraph and are already mentioned above.

The name of each nodeset corresponds to the version of OCP that is deployed by the CRC image.

## Adding new jobs

If a new job needs to be added, it should inherit from `stf-base` ( or one of its child-jobs) which includes common tasks for setting up STF. The new jobs should have minimal configuration lines; either the `scenario` var is passed, which selects a vars file for stf-run-ci, to change its configuration, or the nodeset should be updated, which selects the OCP version.

> Below are examples of how to add a job. Take note of how the `scenario` var and the `nodeset` is passed.

Reviewer 3: ```suggestion
Below in an example of how to add a job. Take note of how the `scenario` var and the `nodeset` is passed.
```

    - job:
        name: stf-crc-nightly_bundles
        parent: stf-base
        abstract: true
        description: |
          Example of a job that extends the `stf-base` job, and passes the `nightly_bundles` scenario var. This job does NOT have a nodeset defined so it must be abstract.
        vars:
          scenario: "nightly_bundles"

> OR

Reviewer 3: ```suggestion
```
This looks more like one complete example than an either/or set of examples.

    - job:
        name: stf-crc-ocp_414-nightly_bundles
        parent: stf-crc-nightly_bundles
        description: |
          Example of a job defining a nodeset to be used.
          Since this job derives from a job with a scenario, it can be run directly.
        nodeset: stf-crc_extracted-ocp414

All non-abstract jobs inheriting from `stf-base` must pass a `scenario` var to work correctly. There is no default value for the `scenario`.
All non-abstract jobs defined in this repo must have a `nodeset` to run correctly. Specifically, the nodeset must include nodes called `controller` and `crc`. This requirements comes from the `stf-base-2node` job.

Once a new job is defined, it should be added to a project or to the `stf-crc-jobs` [template](https://zuul-ci.org/docs/zuul/latest/config/project.html#project-template) in `.zuul.yaml`.
Any job added to a project is run only against changes to that project.
Any job added to the `stf-crc-jobs` project template is run in the other repos across the infrawatch org.

## Troubleshooting

> ## FAQ

Reviewer 2: Perhaps mention what kind of artifacts are generated after a run (logs for controller and crc, and how to navigate through them). Maybe make a list of logs locations that are useful for further debugging.

### How does Zuul work across branches?
Each branch has its own zuul configuration. The configuration for a particular branch lives on that branch.

> To run jobs on a branch, the `.zuul.yaml` file needs to exist on that branch.

Reviewer: Should we mention the filter that is being done here https://review.rdoproject.org/r/c/config/+/51927 that we may need to add the branch to this list ?

Author: This is a general "how does zuul do branches?" question.
We shouldn't need to update the branches to just open a PR.
